cni-repair-controller/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+    runtime::{
+        events::{Event, EventType, Recorder, Reporter},
+        watcher,
+    },
+    Client, Error, Resource, ResourceExt,
+};
+use kubert::Runtime;
+use prometheus_client::{metrics::counter::Counter, registry::Registry};
+use tokio::sync::mpsc::{self, error::TrySendError, Receiver, Sender};
+use tokio::task::JoinHandle;
+
+// ERRNO 95: Operation not supported
+const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+// If the event channel capacity is reached, the event is dropped, but a new one will be emitted
+// in the pod's next crashloop iteration
+const EVENT_CHANNEL_CAPACITY: usize = 32;
+
+const DATA_PLANE_LABEL: &str = "linkerd.io/control-plane-ns";
+const EVENT_ACTION: &str = "Deleting";
+const EVENT_REASON: &str = "LinkerdCNINotConfigured";
+
+#[derive(Clone, Debug)]
+pub struct Metrics {
+    queue_overflow: Counter<u64>,
+    deleted_pods: Counter<u64>,
+}
+
+pub fn run(
+    rt: &mut Runtime,
+    node_name: String,
+    controller_pod_name: String,
+    metrics: Metrics,
+) -> JoinHandle<()> {
+    let pod_evts = rt.watch_all::<Pod>(
+        watcher::Config::default()
+            .labels(DATA_PLANE_LABEL)
+            .fields(&format!("spec.nodeName={node_name}")),
+    );
+    let (tx, rx) = mpsc::channel(EVENT_CHANNEL_CAPACITY);
+    tokio::spawn(process_events(pod_evts, tx, metrics.clone()));
+
+    let client = rt.client();
+    tokio::spawn(process_pods(client, controller_pod_name, rx, metrics))
+}
+
+async fn process_events(
+    pod_evts: impl Stream<Item = watcher::Event<Pod>>,
+    tx: Sender<ObjectReference>,
+    metrics: Metrics,
+) {
+    tokio::pin!(pod_evts);
+    while let Some(evt) = pod_evts.next().await {
+        tracing::trace!(?evt);
+        if let watcher::Event::Applied(pod) = evt {
+            let status = if let Some(ref status) = pod.status {
+                status.clone()
+            } else {
+                tracing::info!("Skipped, no status");
+                continue;
+            };
+
+            let terminated = status
+                .init_container_statuses
+                .unwrap_or_default()
+                .iter()
+                .find(|status| status.name == "linkerd-network-validator")
+                .and_then(|status| status.last_state.as_ref())
+                .and_then(|status| status.terminated.as_ref())
+                .map(|term| term.exit_code == UNSUCCESSFUL_EXIT_CODE)
+                .unwrap_or(false);
+
+            let deleting = pod.metadata.deletion_timestamp.is_some();
+
+            if terminated && !deleting {
+                let namespace = pod.namespace().unwrap();
+                let name = pod.name_any();
+                let object_ref = pod.object_ref(&());
+                // this avoids blocking the event loop
+                match tx.try_send(object_ref) {
+                    Ok(_) => {}
+                    Err(TrySendError::Full(_)) => {
+                        tracing::warn!(%namespace, %name, "Dropped event (channel full)");
```

Comment:
What's the context around these errors? Why is it safe to continue? What should future-you know about this condition? What should users know about this condition?

If understand correctly, we expect to get continuing events for all pods in a bad state? Can we document that here?

With that in mind, is this the appropriate log level and message? If this is a diagnostic for ourselves, perhaps we want to lower this to debug (we have a metric to capture during any diagnostic dump). If this is for the user: what's the proper course of action? Presumably they need to go diagnose why the other task is stuck?

Also consider that, if the queue is full, then we'll probably be in a situation where we log this *many* times (because, presumably, the rate of events is higher than we can process, so we'll remain in that state until the rate decreases). If this is for users we may want to track the states to only log when we enter this state.

With all that in mind, I think it's probably sufficient to lower the log level to debug
```suggestion
                        // If a pod is in a failed state, it should continually be
                        // reported to be in CrashLoopBackoff; so it will naturally
                        // be retried.
                        tracing::debug!(%namespace, %name, "Dropped event (channel full)");
```

Reply:
The comment you added is accurate, and reflects the comment on the `EVENT_CHANNEL_CAPACITY` const. I think we'd hit this failure mostly when the rate of pod deletions is simply too high compared to this channel capacity we've set for ourselves, which I still think is a good idea. I agree in this case the user shouldn't be warned as long as the controller continues processing entries. If on the other hand the failure is caused by latency on kube-api calls in `process_pods`, the additional logs and metrics you suggest below should help diagnosing.